{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PDB program function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def PDB_program():\n",
    "    \"\"\"\n",
    "    This function takes no argument.\\n\n",
    "    After calling this function different options\n",
    "    will be displayed.\\n\n",
    "    A user must read in a file(s) using R option so as to work with\n",
    "    them for other options.\\n\n",
    "    For more descriptions for each option a user can use\n",
    "    H option to display options infromation.\n",
    "    \"\"\"\n",
    "    option_menu = ( \"\"\"\n",
    "    OPTIONS: \n",
    "            READ (R)\n",
    "\n",
    "            SEARCH (S)\n",
    "\n",
    "            WRITE (W)\n",
    "\n",
    "            INFORMATION (I)\n",
    "\n",
    "            ALIGNMENT (A)\n",
    "\n",
    "            HELP (H)\n",
    "\n",
    "            QUIT (Q)\n",
    "    \"\"\")    \n",
    "    print(option_menu)\n",
    "\n",
    "\n",
    "    global option\n",
    "    option = input(\"Choose an option:\") \n",
    "    #option_list = ['R','S','W','I','A','H','Q']\n",
    "    global file_list\n",
    "\n",
    "\n",
    "\n",
    "    while option.lower() in ('r','s','w','i','a','h','q'):\n",
    "        if option.lower() in ('r','s','w','i','a','h','q'):\n",
    "\n",
    "            #if option not in ('R','S','W','I','A','H','Q'):\n",
    "             #   print(\"Invalid option \\nEnter a valid option\")\n",
    "              #  option_list = ['R','S','W','I','A','H','Q']\n",
    "                #if option_list:\n",
    "                    #option = input(\":\")\n",
    "\n",
    "            if option.lower() == 'r':\n",
    "                #global file_list\n",
    "                #file_path = []\n",
    "                file_list = []\n",
    "                import os.path\n",
    "                global file\n",
    "                global path\n",
    "                file = input(\"Enter the name of the PDB file: \")\n",
    "                path = \"/home/enezer/python-mini-project-enezermjema/data/\"\n",
    "                global filepath\n",
    "                filepath = path + file\n",
    "                if os.path.exists(filepath) == True:\n",
    "                    file_list.append(file)\n",
    "                    print(\"File has been successful loaded\")\n",
    "                    #file_path.append(filepath)\n",
    "\n",
    "                    print(option_menu)\n",
    "                    option = input(\"Choose an option:\")\n",
    "\n",
    "                else:\n",
    "                    print(\"The file does not exist \\nPlease enter a valid name\")\n",
    "                    file = input(\"Enter the name of the PDB file: \")\n",
    "\n",
    "            #print(option_menu)\n",
    "            #option = input(\"Choose an option:\")\n",
    "            \n",
    "            \n",
    "            if option.lower() == 's':\n",
    "                import re\n",
    "\n",
    "                print(file_list)\n",
    "                File = input(\"choose a file to work with:\")\n",
    "                file_path = path + File\n",
    "\n",
    "                with open(file_path, \"r\") as output:\n",
    "                   # with open(\"/home/enezer/python-mini-project-enezermjema/data/1dm9.pdb\", 'w') as enezer:\n",
    "                    residues = []\n",
    "                    list1 = []\n",
    "                    for line in output:\n",
    "                        if line.startswith(\"SEQRES\"):\n",
    "                            linesplit = line.split()\n",
    "                            if linesplit[2] == \"A\":\n",
    "                                residues.append(linesplit[4:])\n",
    "                    Amino_dict = {'ALA':'A','CYS':'C','ASP':'D','GLU':'E','PCA':'E','PHE':'F','GLY':'G','HIS':'H','ILE':'I','LYS':'K','LEU':'L','MET':'M',\n",
    "                               'ASN':'N','PRO':'P','GLN':'Q','ARG':'R','SER':'S','THR':'T','VAL':'V','TRP':'W','TYR':'Y','LLP':'K'}\n",
    "\n",
    "                    single_code  = 'TRIMDYIIYRFLLIFGAPTMALFVLTKAYPFLQNYGYAPTLNIPITGSMDTVYSNSTLAPTREEIFYAFTLTSTLCLYYGAWGP'\n",
    "                    for lis_list in residues:                                                    \n",
    "                        for residue in list1:                          \n",
    "                            list1.append(residue)\n",
    "                    #global three_letter\n",
    "                    three_letter = \"\".join(Amino_dict[x] for x in list1)\n",
    "\n",
    "                    pattern = re.compile(r'[GYL]A[PFW][TLVGMT]')\n",
    "                    matches = pattern.finditer(single_code)\n",
    "                    matches_tup = tuple(matches)\n",
    "                    print(\"The number of glycoslation site is %d\"%(len(matches_tup)))\n",
    "\n",
    "                    print(option_menu)\n",
    "                    option = input(\"Choose an option:\")\n",
    "\n",
    "\n",
    "            if option.lower() == 'w':\n",
    "                print(\"\"\"\n",
    "                Options:\n",
    "                        1. Write out coordinate file\n",
    "                        2. Write out sequence (Fasta format) \n",
    "                \"\"\")\n",
    "                write_list = ['1','2']\n",
    "                choice = input(':')\n",
    "\n",
    "                if choice == '1':\n",
    "                    print(file_list)\n",
    "                    File = input(\"choose a file to work with:\")\n",
    "                    file_path = path + File\n",
    "                    new = input(\"Name for the new Coordinate file:\")\n",
    "                    file_path2 = path + new\n",
    "\n",
    "                    print(\"\"\"\n",
    "                    Sub-options:\n",
    "                                a. All atom \n",
    "                                b. Backbone atoms\n",
    "                                c. Alpha carbon atoms only\n",
    "                    \"\"\")\n",
    "                    sub_option = input('Sub-option:')\n",
    "                    sub_list = ['a','b','c']\n",
    "\n",
    "\n",
    "                    if sub_option == 'a':\n",
    "                        with open(file_path, 'r') as amino:\n",
    "                            with open(file_path2, 'w') as atom:\n",
    "                                for line in amino:\n",
    "                                    if line.startswith('ATOM'):\n",
    "                                        atom.writelines(line)\n",
    "\n",
    "                        print(\"File %s has been created successfully\"%new)\n",
    "\n",
    "                        print(option_menu)\n",
    "                        option = input(\"Choose an option:\")\n",
    "\n",
    "                    elif sub_option == 'b':\n",
    "                        with open(file_path, 'r') as amino:\n",
    "                            with open(file_path2, 'w') as backbone:\n",
    "                                for line in amino:\n",
    "                                    if line.startswith(\"ATOM\"):\n",
    "                                        l = line.split()\n",
    "                                        atom = l[2]\n",
    "                                        if atom in [\"CA\",\"C\",\"N\",\"O\"]:\n",
    "                                            ene = print(l[2],l[6],l[7],l[8], sep ='\\t')\n",
    "                                            backbone.writelines(ene)\n",
    "\n",
    "                        print(\"File %s has been created successfully\"%new)\n",
    "\n",
    "                        print(option_menu)\n",
    "                        option = input(\"Choose an option:\")\n",
    "\n",
    "                    elif sub_option == 'c':\n",
    "                        with open(file_path, 'r') as amino:\n",
    "                            with open(file_path2, 'w') as alpha:\n",
    "                                for line in amino:\n",
    "                                    if line.startswith(\"ATOM\"):\n",
    "                                        l = line.split()\n",
    "                                        atom = l[2]\n",
    "                                        if atom in [\"CA\"]:\n",
    "                                            ene = print(l[2],l[6],l[7],l[8], sep ='\\t')\n",
    "                                            alpha.writelines(ene)\n",
    "\n",
    "                        print(\"File %s has been created successfully\"%new)\n",
    "\n",
    "                        print(option_menu)\n",
    "                        option = input(\"Choose an option:\")\n",
    "\n",
    "                elif choice == '2':\n",
    "                    print(file_list)\n",
    "                    File = input(\"choose a file to work with:\")\n",
    "                    file_path = path + File\n",
    "\n",
    "\n",
    "                    print(\"\"\"\n",
    "                    Sub-options:\n",
    "                               1. SEQRES Sequence\n",
    "                               2. Coordinate Sequence\n",
    "                               3. Alignment Sequence \n",
    "                    \"\"\")\n",
    "                    sub_option = input('Sub-option:')\n",
    "                    if sub_option == '1':\n",
    "                        new = input(\"Name for the new SEQRES file:\")\n",
    "                        file_path2 = path + new\n",
    "                        with open(filepath, 'r') as amino:\n",
    "                            with open(file_path2, 'w') as seqres:\n",
    "                                for line in amino:\n",
    "                                    if line.startswith(\"SEQRES\"):\n",
    "                                        linesplit = line.split()\n",
    "                                        if linesplit[2] == \"A\":\n",
    "                                            seqres.writelines(line)\n",
    "\n",
    "                        print(\"File %s has been created successfully\"%new)\n",
    "\n",
    "                        print(option_menu)\n",
    "                        option = input(\"Choose an option:\")\n",
    "\n",
    "                    elif sub_option == '2':\n",
    "                        new = input(\"Name for the new COORDINATE file\")\n",
    "                        file_path2 = path + new\n",
    "                        with open(filepath, 'r') as amino:\n",
    "                            with open(file_path2, 'w') as coordinate:\n",
    "                                for x in amino:\n",
    "                                    if x.startswith(\"ATOM\"):\n",
    "                                        z = x.split()\n",
    "                                        coordinate_index = z[6:10]\n",
    "                                        y = \"\\t\".join(coordinate_index)\n",
    "                                        j = y+'\\n'\n",
    "                                        coordinate.writelines(j)\n",
    "\n",
    "                        print(\"File %s has been created successfully\"%new)\n",
    "\n",
    "                        print(option_menu)\n",
    "                        option = input(\"Choose an option:\")\n",
    "\n",
    "\n",
    "            if option.lower() == 'i':\n",
    "                print(file_list)\n",
    "                File = input(\"choose a file to work with:\")\n",
    "                file_path = path + File\n",
    "\n",
    "                print(\"\"\"\n",
    "                SUB-OPTIONS:\n",
    "                        Display coordinate sequence (1)\n",
    "                        Display SEQRES sequence (2)\n",
    "                        Display Alignment sequence (3)\n",
    "                        Display all non-water ligands in the protein (4)\n",
    "                \"\"\")\n",
    "                choice = input('Enter a Sub-opton:')\n",
    "                if choice == '1':\n",
    "                    with open(filepath, 'r') as amino:\n",
    "                        for x in amino:\n",
    "                            if x.startswith(\"ATOM\"):\n",
    "                                z = x.split()\n",
    "                                coordinate_index = z[6:10]\n",
    "                                y = \"\\t\".join(coordinate_index)\n",
    "                                j = y+'\\n'\n",
    "                                print(j)\n",
    "\n",
    "                                print(option_menu)\n",
    "                                option = input(\"Choose an option:\")\n",
    "\n",
    "                elif choice == '2':\n",
    "                    print(file_list)\n",
    "                    File = input(\"choose a file to work with:\")\n",
    "                    file_path = path + File\n",
    "                    with open(filepath, 'r') as amino:\n",
    "                        for line in amino:\n",
    "                            if line.startswith(\"SEQRES\"):\n",
    "                                linesplit = line.split()\n",
    "                                if linesplit[2] == \"A\":\n",
    "                                    print(line)\n",
    "\n",
    "                                    print(option_menu)\n",
    "                                    option = input(\"Choose an option:\")\n",
    "\n",
    "\n",
    "            if option.lower() == 'a':\n",
    "                print(file_list)\n",
    "                #print(file_path)\n",
    "                file1 = input(\"choose a file to work with:\")\n",
    "                file2 = input(\"choose a file to work with:\")\n",
    "                filepath1 = path + file1\n",
    "                filepath2 = path + file2\n",
    "                new = input(\"Name for the Alignment file:\")\n",
    "                file_path3 = path + new\n",
    "\n",
    "\n",
    "                from Bio import pairwise2\n",
    "                from Bio.pairwise2 import format_alignment\n",
    "\n",
    "                with open(path + file1, \"r\") as pdb1:\n",
    "                    with open(path + file2, \"r\") as pdb2:\n",
    "                        with open(file_path3, 'w') as pdb3:\n",
    "                            for line in pdb1:\n",
    "                                if line.startswith(\"SEQRES\"):\n",
    "                                    linesplit = line.split()\n",
    "                                    if linesplit[2] == \"A\":\n",
    "                                        global X\n",
    "                                        X = (line[18:])\n",
    "                                        #print(X)\n",
    "                            for line in pdb2:\n",
    "                                if line.startswith(\"SEQRES\"):\n",
    "                                    linesplit = line.split()\n",
    "                                    if linesplit[2] == \"A\":\n",
    "                                        global Y\n",
    "                                        Y = (line[18:])\n",
    "                                        #print(Y)\n",
    "                            alignments = pairwise2.align.globalxx(X, Y)\n",
    "                            for a in alignments:\n",
    "                                print(format_alignment(*a))\n",
    "                            stri = ', '.join(map(str, alignments))\n",
    "                            pdb3.writelines(stri)\n",
    "\n",
    "                    #for a in alignments:\n",
    "                     #   aligned = print(format_alignment(*a))\n",
    "                      #  pdb3.writelines(aligned)\n",
    "\n",
    "\n",
    "                            print(option_menu)\n",
    "                            option = input(\"Choose an option:\")\n",
    "\n",
    "            if option.lower() == 'h':\n",
    "                print(\"\"\"\n",
    "\n",
    "                THE FOLLOWING ARE OPTIONS IN THE PROGRAM WITH THEIR SHORT DESCRIPTION.\n",
    "\n",
    "                        READ (R) - This option load in a PDB file that is present in the data diretory.\n",
    "\n",
    "                        SEARCH (S) - Gives a list of PDB files already loaded in the system.\\n\n",
    "                                     It gives the number of glycoslation sites present in the\n",
    "                                     file choosen and display sequence with glycoslation motifs \n",
    "                                     shown in lower case and the rest in upper case.\n",
    "\n",
    "                        WRITE (W) -  This option write out a file of a given name by the user depending\n",
    "                                     on the following sub-option chosen;\\n\n",
    "                                              1. Write out coordinate file \n",
    "                                                   a. All atom \n",
    "                                                   b. Backbone atoms \n",
    "                                                   c. Alpha carbon atoms only \n",
    "                                              2.Write out sequence (Fasta format) \n",
    "                                                   1. SEQRES Sequence\n",
    "                                                   2. Coordinate Sequence\n",
    "                                                   3. Alignment Sequence \n",
    "\n",
    "                        INFORMATION (I) - This option has sub-options for:\\n\n",
    "                                            1. Displaying coordinate sequence\n",
    "                                            2. Displaying SEQRES sequence \n",
    "                                            3. Displaying Alignment sequence\n",
    "                                            4. Displaying all non-water ligands in the protein\n",
    "\n",
    "                        ALIGNMENT (A) - Creates and write to a file an alignment of two PDB sequence \n",
    "                                        from two chosen files.\\n\n",
    "\n",
    "                        HELP (H) - This option display all the options in the program with their description.\\n\n",
    "\n",
    "                        QUIT (Q) - Terminate the program.\\n\n",
    "\n",
    "                        THANK YOU...\n",
    "\n",
    "                        \"\"\")\n",
    "\n",
    "                print(option_menu)\n",
    "                option = input(\"Choose an option:\")\n",
    "\n",
    "            if option.lower() == 'q': \n",
    "                print(\n",
    "                \"\"\"\n",
    "                PROGRAM TERMINATED\\n\n",
    "                BYE...\n",
    "                \"\"\")\n",
    "                break\n",
    "\n",
    "\n",
    "        else:\n",
    "            print(\"Invalid option \\nEnter a valid option\")\n",
    "            option = input(\"Choose an option:\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    OPTIONS: \n",
      "            READ (R)\n",
      "\n",
      "            SEARCH (S)\n",
      "\n",
      "            WRITE (W)\n",
      "\n",
      "            INFORMATION (I)\n",
      "\n",
      "            ALIGNMENT (A)\n",
      "\n",
      "            HELP (H)\n",
      "\n",
      "            QUIT (Q)\n",
      "    \n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Choose an option: q\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "                PROGRAM TERMINATED\n",
      "\n",
      "                BYE...\n",
      "                \n"
     ]
    }
   ],
   "source": [
    "PDB_program()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
